### ISSUE TITLE
Possibly wrong implementation of self-attention module?

### ISSUE THREAD

**Issue by richardwth** [taki0112/Self-Attention-GAN-Tensorflow#3]
I think your implementation is wrong. The tensor shapes were messed up:

`f = conv(x, ch // 8, kernel=1, stride=1, scope='f_conv')   # [bs H W C//8]`
`g = conv(x, ch // 8, kernel=1, stride=1, scope='g_conv')   # [bs H W C//8]`
`h = conv(x, ch, kernel=1, stride=1, scope='h_conv')   # [bs H W C]`
`s = tf.matmul(g, f, transpose_b=True)   # [bs H W W]`
`attention_shape = s.shape`
`s = tf.reshape(s, shape=[attention_shape[0], -1, attention_shape[-1]])  # [bs, HW, W]`
`beta = tf.nn.softmax(s, axis=1)  # [bs, HW, W]`
`beta = tf.reshape(beta, shape=attention_shape)   # bs H W W`
`o = tf.matmul(beta, h)   # [bs H W C]`

The self-attention module is essentially a non-local module from:
X. Wang, R. Girshick, A. Gupta, and K. He. Non-local neural networks. In CVPR, 2018.
Based on Figure 2 of above paper, `attention_shape` should be `[bs, H*W, H*W]`.

**Comment by taki0112** [comment]
Thank you
I fix it

**Comment by richardwth** [comment]
I implemented the self-attention myself, and tried to compare the results against that without. So far self-attention did not improve the model performance on cifar-10 (for all learning rates I tried). In my implementation, batch normalization / spectral normalization are also used inside the self-attention module. Perhaps you should also consider this (as if there is no normalization in self-attention layer and the gamma parameter, the discriminator will not be Lipchitz anymore.)

I haven't tried on celebA because self-attention is quite computationally intensive. I am working on improving it. :D
